File: src/windows-native/LanMessenger/Core/Networking/NetworkInterfaceMonitor.cs
using LanMessenger.Core.Services;
using System.Net;
using System.Net.NetworkInformation;
using System.Net.Sockets;

namespace LanMessenger.Core.Networking;

// One usable IPv4 interface on this machine. Snapshot value — never mutated
// after construction; the monitor publishes a fresh set on every change.
public sealed record NetworkAdapterSnapshot(
    string    InterfaceId,
    string    Description,
    IPAddress LocalIP,
    IPAddress SubnetMask,
    IPAddress BroadcastAddress);

// Tracks the set of IPv4 interfaces that are eligible for LAN discovery.
//
// "Eligible" means:
//   - OperationalStatus is Up
//   - Has at least one non-loopback IPv4 unicast address
//   - The address is not in the APIPA link-local range (169.254/16) — those
//     indicate "no real network" on Windows when DHCP fails, and shipping
//     beacons there is pointless
//
// The monitor publishes a snapshot of the current adapter set and fires
// Changed whenever the set differs from the previous snapshot. It listens to
// both NetworkChange.NetworkAddressChanged (fires when an IP comes/goes) and
// NetworkChange.NetworkAvailabilityChanged (fires when the system flips
// between "any network" and "no network"), and also polls every 5 s as a
// safety net — those events are documented as best-effort on Windows and
// occasionally miss adapter transitions (especially Wi-Fi roam, VPN bring-up,
// and Hyper-V/WSL adapter creation).
//
// IsLocalNetworkAvailable is true whenever the snapshot is non-empty. This is
// what the UI should treat as "the app has a usable local network", not
// "internet is reachable" — LAN messaging doesn't need internet.
public sealed class NetworkInterfaceMonitor : IDisposable
{
    public event Action? Changed;

    public IReadOnlyList<NetworkAdapterSnapshot> Adapters { get; private set; } = [];
    public bool IsLocalNetworkAvailable => Adapters.Count > 0;

    public HashSet<string> LocalIPs =>
        [.. Adapters.Select(a => a.LocalIP.ToString())];

    private System.Threading.Timer? _pollTimer;
    private readonly object _lock = new();
    private bool _started;

    public void Start()
    {
        lock (_lock)
        {
            if (_started) return;
            _started = true;
        }

        NetworkChange.NetworkAddressChanged      += OnNetworkChanged;
        NetworkChange.NetworkAvailabilityChanged += OnAvailabilityChanged;

        // Initial snapshot synchronously so callers see populated state immediately.
        Refresh(reason: "initial");

        // 5 s safety-net poll — the OS events are documented as best-effort
        // and occasionally miss transitions on Wi-Fi roams or VPN bring-up.
        _pollTimer = new System.Threading.Timer(_ => Refresh(reason: "poll"), null,
            TimeSpan.FromSeconds(5), TimeSpan.FromSeconds(5));
    }

    public void Stop()
    {
        lock (_lock)
        {
            if (!_started) return;
            _started = false;
        }

        NetworkChange.NetworkAddressChanged      -= OnNetworkChanged;
        NetworkChange.NetworkAvailabilityChanged -= OnAvailabilityChanged;
        _pollTimer?.Dispose();
        _pollTimer = null;
    }

    public void Dispose() => Stop();

    private void OnNetworkChanged(object? sender, EventArgs e)      => Refresh(reason: "addr-changed");
    private void OnAvailabilityChanged(object? sender, NetworkAvailabilityEventArgs e)
        => Refresh(reason: $"avail-changed={e.IsAvailable}");

    private void Refresh(string reason)
    {
        IReadOnlyList<NetworkAdapterSnapshot> fresh;
        try { fresh = Enumerate(); }
        catch (Exception ex)
        {
            LanLogger.Warn("NetMonitor", $"enumerate failed ({reason}): {ex.GetType().Name} {ex.Message}");
            return;
        }

        var prev = Adapters;
        if (AdapterSetsEqual(prev, fresh)) return;

        Adapters = fresh;
        LanLogger.Info("NetMonitor",
            $"interfaces changed ({reason}): " +
            $"was={prev.Count} now={fresh.Count} " +
            $"ips=[{string.Join(",", fresh.Select(a => a.LocalIP))}] " +
            $"available={IsLocalNetworkAvailable}");
        try { Changed?.Invoke(); }
        catch (Exception ex)
        {
            LanLogger.Warn("NetMonitor", $"Changed handler threw: {ex.GetType().Name} {ex.Message}");
        }
    }

    // MARK: - Enumeration

    private static IReadOnlyList<NetworkAdapterSnapshot> Enumerate()
    {
        var list = new List<NetworkAdapterSnapshot>();
        foreach (var iface in NetworkInterface.GetAllNetworkInterfaces())
        {
            if (iface.OperationalStatus != OperationalStatus.Up) continue;
            if (iface.NetworkInterfaceType == NetworkInterfaceType.Loopback) continue;

            // Tunnel and PPP can be legitimate (Wireguard, modem) — keep them.
            // Filtering is intentionally permissive: it's safer to send a beacon to a
            // dead VPN adapter than to silently exclude a working LAN adapter.

            var props = iface.GetIPProperties();
            foreach (var uni in props.UnicastAddresses)
            {
                if (uni.Address.AddressFamily != AddressFamily.InterNetwork) continue;

                var ip = uni.Address;
                if (IPAddress.IsLoopback(ip)) continue;

                var bytes = ip.GetAddressBytes();
                // 169.254/16 (APIPA) — "DHCP failed", no real network. Skip so we
                // don't pollute beacons with addresses no peer can reach.
                if (bytes[0] == 169 && bytes[1] == 254) continue;
                // 0.0.0.0 — unbound interface, skip.
                if (bytes is [0, 0, 0, 0]) continue;

                IPAddress mask;
                try { mask = uni.IPv4Mask ?? IPAddress.Parse("255.255.255.0"); }
                catch { mask = IPAddress.Parse("255.255.255.0"); }

                // Pre-compute the IPv4 directed-broadcast address (host bits all 1).
                // Used as one of the beacon targets so peers on the same subnet
                // receive even when limited-broadcast (255.255.255.255) is dropped
                // by some routers/switches.
                var broadcast = ComputeBroadcast(ip, mask);

                list.Add(new NetworkAdapterSnapshot(
                    InterfaceId:      iface.Id,
                    Description:      iface.Description,
                    LocalIP:          ip,
                    SubnetMask:       mask,
                    BroadcastAddress: broadcast));
            }
        }
        return list;
    }

    private static IPAddress ComputeBroadcast(IPAddress ip, IPAddress mask)
    {
        var ipBytes   = ip.GetAddressBytes();
        var maskBytes = mask.GetAddressBytes();
        if (ipBytes.Length != 4 || maskBytes.Length != 4) return IPAddress.Broadcast;
        var bcast = new byte[4];
        for (var i = 0; i < 4; i++) bcast[i] = (byte)(ipBytes[i] | ~maskBytes[i]);
        return new IPAddress(bcast);
    }

    private static bool AdapterSetsEqual(
        IReadOnlyList<NetworkAdapterSnapshot> a,
        IReadOnlyList<NetworkAdapterSnapshot> b)
    {
        if (a.Count != b.Count) return false;
        var aSet = a.Select(x => $"{x.InterfaceId}|{x.LocalIP}|{x.SubnetMask}").ToHashSet();
        return b.All(x => aSet.Contains($"{x.InterfaceId}|{x.LocalIP}|{x.SubnetMask}"));
    }
}

File: src/macos/LanMessenger/Core/Networking/NetworkInterfaceMonitor.swift
import Foundation
import Network
import Darwin

// One usable IPv4 interface on this machine. Snapshot value — never mutated
// after construction; the monitor publishes a fresh set on every change.
struct NetworkAdapterSnapshot: Equatable, Hashable {
    let name: String              // BSD interface name, e.g. en0, en1, utun3
    let localIP: String           // "192.168.1.42"
    let subnetMask: String        // "255.255.255.0"
    let broadcastAddress: String  // "192.168.1.255"
}

// Tracks the set of IPv4 interfaces eligible for LAN discovery on macOS.
//
// "Eligible" means:
//   - IFF_UP set, IFF_LOOPBACK clear
//   - Has at least one IPv4 unicast address
//   - The address is not in the APIPA link-local range (169.254/16) — those
//     indicate "no real network" when DHCP fails on macOS
//
// Publishes a snapshot of the current adapter set and fires `onChange`
// whenever the set differs from the previous one. Listens to NWPathMonitor
// for OS-level network transitions (Wi-Fi reconnect, sleep/resume, VPN
// toggle) and runs a 5-second polling safety net.
//
// IsLocalNetworkAvailable is true whenever the snapshot is non-empty. The UI
// should treat that as "the app has a usable local network", not "internet is
// reachable" — LAN messaging doesn't need internet.
final class NetworkInterfaceMonitor {

    // Observers receive notifications on the main queue whenever the adapter
    // set changes. Returned token is opaque; pass it to removeObserver to stop.
    typealias Observer = () -> Void

    private(set) var adapters: [NetworkAdapterSnapshot] = []

    var isLocalNetworkAvailable: Bool { !adapters.isEmpty }

    var localIPs: Set<String> { Set(adapters.map { $0.localIP }) }

    private let pathMonitor = NWPathMonitor()
    private let pathQueue = DispatchQueue(label: "com.dave.lanmessenger.netmonitor", qos: .utility)
    private var pollTimer: DispatchSourceTimer?
    private var started = false

    private var observers: [(id: UUID, fn: Observer)] = []
    private let observerLock = NSLock()

    @discardableResult
    func addObserver(_ fn: @escaping Observer) -> UUID {
        observerLock.lock(); defer { observerLock.unlock() }
        let id = UUID()
        observers.append((id, fn))
        return id
    }

    func removeObserver(_ id: UUID) {
        observerLock.lock(); defer { observerLock.unlock() }
        observers.removeAll { $0.id == id }
    }

    private func notifyObservers() {
        observerLock.lock()
        let snapshot = observers
        observerLock.unlock()
        DispatchQueue.main.async {
            for (_, fn) in snapshot { fn() }
        }
    }

    func start() {
        guard !started else { return }
        started = true

        // Initial snapshot synchronously so callers see populated state immediately.
        refresh(reason: "initial")

        pathMonitor.pathUpdateHandler = { [weak self] _ in
            self?.refresh(reason: "nwpath")
        }
        pathMonitor.start(queue: pathQueue)

        // 5 s safety-net poll — NWPathMonitor occasionally misses transitions
        // on Wi-Fi roams and VPN bring-up; this catches the strays.
        let timer = DispatchSource.makeTimerSource(queue: pathQueue)
        timer.schedule(deadline: .now() + .seconds(5), repeating: .seconds(5))
        timer.setEventHandler { [weak self] in self?.refresh(reason: "poll") }
        timer.resume()
        pollTimer = timer
    }

    func stop() {
        guard started else { return }
        started = false
        pollTimer?.cancel()
        pollTimer = nil
        pathMonitor.cancel()
    }

    // MARK: - Refresh

    private let refreshLock = NSLock()

    private func refresh(reason: String) {
        refreshLock.lock()
        let fresh = Self.enumerate()
        let prev = adapters
        let equal = fresh.count == prev.count &&
            Set(fresh).isSubset(of: Set(prev)) &&
            Set(prev).isSubset(of: Set(fresh))
        if !equal { adapters = fresh }
        refreshLock.unlock()

        if equal { return }
        NetLogger.info("NetMonitor",
            "interfaces changed (\(reason)): was=\(prev.count) now=\(fresh.count) " +
            "ips=[\(fresh.map { $0.localIP }.joined(separator: ","))] " +
            "available=\(isLocalNetworkAvailable)")
        notifyObservers()
    }

    // MARK: - Interface enumeration

    static func enumerate() -> [NetworkAdapterSnapshot] {
        var result: [NetworkAdapterSnapshot] = []

        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let first = ifaddr else { return [] }
        defer { freeifaddrs(ifaddr) }

        var ptr: UnsafeMutablePointer<ifaddrs>? = first
        while let current = ptr {
            defer { ptr = current.pointee.ifa_next }

            let flags = Int32(current.pointee.ifa_flags)
            guard (flags & IFF_UP) != 0,
                  (flags & IFF_LOOPBACK) == 0,
                  let addrPtr = current.pointee.ifa_addr,
                  addrPtr.pointee.sa_family == UInt8(AF_INET) else { continue }

            let name = current.pointee.ifa_name.map { String(cString: $0) } ?? ""

            let localIP = Self.numericHost(addrPtr)
            guard !localIP.isEmpty,
                  !localIP.hasPrefix("169.254."),
                  localIP != "0.0.0.0" else { continue }

            let mask: String = {
                guard let maskPtr = current.pointee.ifa_netmask else { return "255.255.255.0" }
                let m = Self.numericHost(maskPtr)
                return m.isEmpty ? "255.255.255.0" : m
            }()

            let broadcast = Self.computeBroadcast(ip: localIP, mask: mask)

            result.append(NetworkAdapterSnapshot(
                name: name,
                localIP: localIP,
                subnetMask: mask,
                broadcastAddress: broadcast))
        }
        return result
    }

    private static func numericHost(_ addr: UnsafePointer<sockaddr>) -> String {
        var host = [CChar](repeating: 0, count: Int(NI_MAXHOST))
        let len = socklen_t(addr.pointee.sa_len)
        let rc = getnameinfo(addr, len, &host, socklen_t(host.count), nil, 0, NI_NUMERICHOST)
        return rc == 0 ? String(cString: host) : ""
    }

    private static func computeBroadcast(ip: String, mask: String) -> String {
        let ipParts = ip.split(separator: ".").compactMap { UInt8($0) }
        let mkParts = mask.split(separator: ".").compactMap { UInt8($0) }
        guard ipParts.count == 4, mkParts.count == 4 else { return "255.255.255.255" }
        var b = [UInt8](repeating: 0, count: 4)
        for i in 0..<4 { b[i] = ipParts[i] | ~mkParts[i] }
        return "\(b[0]).\(b[1]).\(b[2]).\(b[3])"
    }
}
